src/prompt_siren/sandbox_managers/docker/plugins/abstract.py

```diff
@@ -0,0 +1,235 @@
+"""Abstract Docker client interface for supporting multiple execution backends.
+
+This module provides abstractions for Docker operations that can be implemented
+by different backends.
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import AsyncIterator
+from typing import Any
+
+from ...abstract import ExecOutput
+
+
+class AbstractContainer(ABC):
+    """Abstract interface for Docker containers."""
+
+    @abstractmethod
+    async def start(self) -> None:
+        """Start the container."""
+        ...
+
+    @abstractmethod
+    async def stop(self) -> None:
+        """Stop the container."""
+        ...
+
+    @abstractmethod
+    async def delete(self) -> None:
+        """Delete the container."""
+        ...
+
+    @abstractmethod
+    async def show(self) -> dict[str, Any]:
+        """Get container details.
+
+        Returns:
+            Dict with container information including Id, State, Config, HostConfig
+        """
+        ...
+
+    @abstractmethod
+    async def exec(
+        self,
+        cmd: list[str],
+        stdin: str | bytes | None,
+        user: str,
+        environment: dict[str, str] | None,
+        workdir: str | None,
+        timeout: int,
+    ) -> ExecOutput:
+        """Execute a command in the container.
+
+        Args:
+            cmd: Command to execute (already in bash -c format)
+            stdin: Optional stdin data to pass to the command
+            user: User to run as
+            environment: Environment variables
+            workdir: Working directory
+            timeout: Timeout in seconds
+
+        Returns:
+            ExecOutput containing stdout/stderr chunks and exit code
+        """
+        ...
+
+    @abstractmethod
+    async def log(self, stdout: bool, stderr: bool) -> list[str]:
+        """Get container logs.
+
+        Args:
+            stdout: Include stdout
+            stderr: Include stderr
+
+        Returns:
+            List of log lines
+        """
+        ...
+
+    @abstractmethod
+    async def commit(self, repository: str, tag: str) -> None:
+        """Commit container to an image.
+
+        Args:
+            repository: Repository name
+            tag: Image tag
+        """
+        ...
+
+
+class AbstractNetwork(ABC):
+    """Abstract interface for Docker networks."""
+
+    @abstractmethod
+    async def show(self) -> dict[str, Any]:
+        """Get network details.
+
+        Returns:
+            Dict with network information including Id, Driver, Internal
+        """
+        ...
+
+    @abstractmethod
+    async def delete(self) -> None:
+        """Delete the network."""
+        ...
+
+
+class AbstractDockerClient(ABC):
```

Comment:
We can also make this generic of the container and network types for better type safety